src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/TraceEventExtensions.cs

```diff
@@ -57,26 +58,269 @@ public static bool TryGetCounterPayload(this TraceEvent traceEvent, CounterFilte
+                }
+                else if (traceEvent.EventName == "HistogramLimitReached")
+                {
+                    HandleHistogramLimitReached(traceEvent, sessionId, out individualPayload);
+                }
+                else if (traceEvent.EventName == "Error")
+                {
+                    HandleError(traceEvent, sessionId, out individualPayload);
+                }
+                else if (traceEvent.EventName == "ObservableInstrumentCallbackError")
+                {
+                    HandleObservableInstrumentCallbackError(traceEvent, sessionId, out individualPayload);
+                }
+                else if (traceEvent.EventName == "MultipleSessionsNotSupportedError")
+                {
+                    HandleMultipleSessionsNotSupportedError(traceEvent, sessionId, out individualPayload);
+                }
+
+                if (null != individualPayload)
+                {
+                    payload.Add(individualPayload);
+                }
+
+                return null != payload && payload.Any();
+            }
+
             return false;
         }
 
+        public static bool TryGetIndividualCounterPayload(this TraceEvent traceEvent, CounterFilter filter, out ICounterPayload payload)
+        {
+            bool gotCounterPayload = TryGetCounterPayload(traceEvent, filter, null, out List<ICounterPayload> payloadsList);
+
+            payload = payloadsList.FirstOrDefault();
+
+            return gotCounterPayload;
+        }
+
+        private static void HandleGauge(TraceEvent obj, CounterFilter filter, string sessionId, out ICounterPayload payload)
+        {
+            payload = null;
+
+            string payloadSessionId = (string)obj.PayloadValue(0);
+
+            if (payloadSessionId != sessionId)
+            {
+                return;
+            }
+
+            string meterName = (string)obj.PayloadValue(1);
+            //string meterVersion = (string)obj.PayloadValue(2);
+            string instrumentName = (string)obj.PayloadValue(3);
+            string unit = (string)obj.PayloadValue(4);
+            string tags = (string)obj.PayloadValue(5);
+            string lastValueText = (string)obj.PayloadValue(6);
+
+            if (!filter.IsIncluded(meterName, instrumentName))
+            {
+                return;
+            }
+
+            // the value might be an empty string indicating no measurement was provided this collection interval
+            if (double.TryParse(lastValueText, out double lastValue))
+            {
+                payload = new GaugePayload(meterName, instrumentName, null, unit, tags, lastValue, obj.TimeStamp);
+            }
+            else
+            {
+                // for observable instruments we assume the lack of data is meaningful and remove it from the UI
+                // this happens when the Gauge callback function throws an exception.
+
+                //TODO Can this occur for other meter types?
```

Comment:
I'd expect having lastValueText="" can happen for various reasons, but I believe Gauge and Histogram are the only ones where we might get an empty value in one interval after we had already saw non-empty values previously. For histogram it just means there were no data points recorded during that period which is a normal occurrence. For Gauge we expect the callback function to be providing a value every interval even if it is unchanged. If a Gauge stops providing a value presumably that means the thing it is measuring no longer exists. 

Hopefully most of the behavior of MetricsEventSource should be visible in the MetricEventSource unit tests: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs
Test for output when data isn't provided.
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs#L313